src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -0,0 +1,115 @@
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.json.simple.JSONObject;
+
+/** Servlet that stores and shows images and comments. */
+@WebServlet("/search")
+public class SearchServlet extends HttpServlet {
+
+  @Override
+  /* Show classes. */
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    List<String> classNames = new ArrayList<>();
+    List<String> professorNames = new ArrayList<>();
+    List<Filter> filters = new ArrayList<>();
+    if (!request.getParameter("className").isEmpty()) {
+      String className = request.getParameter("className");
+      Filter classFilter = new FilterPredicate("Name", FilterOperator.EQUAL, className);
+      filters.add(classFilter);
+    }
+    if (!request.getParameter("profName").isEmpty()) {
+      String profName = request.getParameter("profName");
+      Filter profFilter = new FilterPredicate("Professor", FilterOperator.EQUAL, profName);
+      filters.add(profFilter);
+    }
+    List<Integer> units = new ArrayList<>();
+    if (!request.getParameter("units").isEmpty()) {
+      List<String> strUnits = Arrays.asList(request.getParameter("units").split(","));
+      for (String number : strUnits) {
+        units.add(Integer.valueOf(number));
+      }
+      if (units.size() != 0) {
+        Filter unitsFilter = new FilterPredicate("Units", FilterOperator.IN, units);
+        filters.add(unitsFilter);
+      }
+    }
+
+    Query q = new Query("Class");
+    // Add filters depending on number of fields specified
+    if (filters.size() == 1) {
+      q.setFilter(filters.get(0));
+    } else if (filters.size() == 2) {
+      q.setFilter(CompositeFilterOperator.and(filters.get(0), filters.get(1)));
+    } else if (filters.size() == 3) {
+      q.setFilter(
+          CompositeFilterOperator.and(
+              CompositeFilterOperator.and(filters.get(0), filters.get(1), filters.get(2))));
+    }
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    List<Entity> classes = datastore.prepare(q).asList(FetchOptions.Builder.withDefaults());
+    for (Entity entity : classes) {
+      String name = (String) entity.getProperty("Name");
+      classNames.add(name);
+      String professor = (String) entity.getProperty("Professor");
+      professorNames.add(professor);
+    }
+
+    // Generate links for class pages and send them back - set variables in URL
+    String classNameJson = new Gson().toJson(classNames);
+    String profNameJson = new Gson().toJson(professorNames);
+    JSONObject obj = new JSONObject();
+    obj.put("classNames", classNameJson);
```

Comment:
Since the class name and professor name are directly related I think it make more sense to package them together as an object and then return a list of those objects. As it is now you have to form that relationship on the client side by matching indexes, not ideal.

Reply:
Ah makes sense, if I make a class for this now it'll also allow me to add more/change attributes as we start including more details in the search filter

Reply:
A class could be a good idea but keep in mind that you'll have to convert that class to a JSON object to send to the client side

Reply:
Yeah - I think I got that up and running, it ended up working out pretty simply